src/functional/iterator/transforms/inline_lifts.py

```diff
@@ -1,99 +1,214 @@
+import dataclasses
 from collections.abc import Callable
 from typing import Optional
 
-from eve import NodeTranslator
+import eve
+from eve import NodeTranslator, traits
+from functional.ffront import itir_makers as im
 from functional.iterator import ir
+from functional.iterator.transforms.inline_lambdas import inline_lambda
 
 
-class InlineLifts(NodeTranslator):
+def _generate_unique_symbol(
+    desired_name: Optional[eve.SymbolName | tuple[ir.Lambda | ir.SymRef, int]] = None,
+    occupied_names=None,
+    occupied_symbols=None,
+):
+    occupied_names = occupied_names or set()
+    occupied_symbols = occupied_symbols or set()
+    if not desired_name:
+        desired_name = "__sym"
+    elif isinstance(desired_name, tuple):
+        fun, arg_idx = desired_name
+        if isinstance(fun, ir.Lambda):
+            desired_name = fun.params[arg_idx].id
+        else:
+            desired_name = f"__arg{arg_idx}"
+
+    new_symbol = ir.Sym(id=desired_name)
+    # make unique
+    while new_symbol.id in occupied_names or new_symbol in occupied_symbols:
+        new_symbol = ir.Sym(id=new_symbol.id + "_")
+    return new_symbol
+
+
+def _is_lift(node: ir.Node):
+    return (
+        isinstance(node, ir.FunCall)
+        and isinstance(node.fun, ir.FunCall)
+        and node.fun.fun == ir.SymRef(id="lift")
+    )
+
+
+def _is_shift_lift(node: ir.Expr):
+    return (
+        isinstance(node, ir.FunCall)
+        and isinstance(node.fun, ir.FunCall)
+        and node.fun.fun == ir.SymRef(id="shift")
+        and isinstance(node.args[0], ir.FunCall)
+        and isinstance(node.args[0].fun, ir.FunCall)
+        and node.args[0].fun.fun == ir.SymRef(id="lift")
+    )
+
+
+def _is_scan(node: ir.FunCall):
+    return node.fun == ir.SymRef(id="scan")
+
+
+def _transform_and_extract_lift_args(
+    node: ir.FunCall,
+    symtable: dict[eve.SymbolName, ir.Sym],
+    extracted_args: dict[ir.Sym, ir.Expr],
+):
+    """
+    Transform and extract non-symbol arguments of a lifted stencil call.
+
+    E.g. ``lift(lambda a: ...)(sym1, expr1)`` is transformed into
+    ``lift(lambda a: ...)(sym1, sym2)`` with the extracted arguments
+    being ``{sym1: sym1, sym2: expr1}``.
+    """
+    assert _is_lift(node)
+    inner_stencil = node.fun.args[0]
+
+    new_args = []
+    for i, arg in enumerate(node.args):
+        if isinstance(arg, ir.SymRef):
+            sym = ir.Sym(id=arg.id)
+            assert sym not in extracted_args or extracted_args[sym] == arg
+            extracted_args[sym] = arg
+            new_args.append(arg)
+        else:
+            new_symbol = _generate_unique_symbol(
+                desired_name=(inner_stencil, i),
+                occupied_names=symtable.keys(),
+                occupied_symbols=extracted_args.keys(),
+            )
+            # assert new_symbol not in extracted_args
```

Comment:
Why commented out? (And if correctly commented out, please remove or add explanation)